File: java/src/main/java/toolgood/algorithm/internals/functions/mathsum2/Function_NORMSINV.java
package toolgood.algorithm.internals.functions.mathsum2;

import java.math.BigDecimal;
import java.util.List;

import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.internals.ParameterType;
import toolgood.algorithm.internals.functions.FunctionBase;
import toolgood.algorithm.internals.functions.Function_1;
import toolgood.algorithm.internals.functions.NoneEngine;
import toolgood.algorithm.mathNet.ExcelFunctions;

public final class Function_NORMSINV extends Function_1 {
    public Function_NORMSINV(FunctionBase func1) {
        super(func1);
    }

    @Override
    public String Name() {
        return "NormSInv";
    }

    @Override
    public Operand Evaluate(AlgorithmEngine engine, java.util.function.BiFunction<AlgorithmEngine, String, Operand> tempParameter) {
        Operand args1 = GetNumber_1(engine, tempParameter);
        if (args1.IsErrorOrNone()) {
            return args1;
        }

        double probability = args1.DoubleValue();
        if (probability <= 0.0 || probability >= 1.0) {
            return Operand.Error("Function '{0}' parameter is error!", "NormSInv");
        }
        return Operand.Create(ExcelFunctions.NormSInv(probability));
    }

    @Override
    public OperandType GetResultType() {
        return OperandType.NUMBER;
    }

    @Override
    public void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, String op, String val) {
        func1.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
    }
}

File: JavaScript/src/Internals/Functions/MathSum2/Function_NORMSINV.js
import { Function_1 } from '../Function_1.js';
import { Operand } from '../../../Operand.js';
import { ExcelFunctions } from '../../../MathNet/ExcelFunctions.js';

class Function_NORMSINV extends Function_1 {
    get Name() {
        return "NormSInv";
    }

    constructor(a) {
        super(a);
    }

    evaluate(engine, tempParameter) {
        let args1 = this.getNumber_1(engine, tempParameter);
        if (args1.IsError) return args1;
        let p = args1.DoubleValue;
        if (p <= 0.0 || p >= 1.0) {
            return this.parameterError(1);
        }
        return Operand.Create(ExcelFunctions.NormSInv(p));
    }
}

export { Function_NORMSINV };
